src/Match.ts

```diff
@@ -6,13 +6,19 @@ export type Match<
   matcher: UnnamedMatch<T, Return> | MatchesWithUnderscore<T, Return> | NamedMatches<Union.ListOf<T>, Return>
 ) => (t: T) => Return
 
-type NamedMatches<T extends ReadonlyArray<{__type: string}>, Return> = 
-  T[1] extends undefined ? NamedMatch<T[0], Return> : NamedMatch<T[0], Return> & NamedMatches<List.Tail<T>, Return>
-
 type MatchesWithUnderscore<All extends {__type: string}, Return> = UnnamedMatch<All, Return> & NamedMatch<All, Return>
 
 type UnnamedMatch<Param, Return> = 
   { _: (match: Param) => Return }
 
+type NamedMatches<T extends ReadonlyArray<{__type: string}>, Return> = 
```

Comment:
As I dont recall was this change is supposed to do, what are the errors that arise if you revert it ?